Code/Engine/GameEngine/StateMachine/Implementation/StateMachineBuiltins.cpp

```diff
@@ -0,0 +1,184 @@
+#include <GameEngine/GameEnginePCH.h>
+
+#include <Foundation/IO/TypeVersionContext.h>
+#include <GameEngine/StateMachine/Implementation/StateMachineBuiltins.h>
+
+// clang-format off
+EZ_BEGIN_DYNAMIC_REFLECTED_TYPE(ezStateMachineState_NestedStateMachine, 1, ezRTTIDefaultAllocator<ezStateMachineState_NestedStateMachine>)
+{
+  EZ_BEGIN_PROPERTIES
+  {
+    EZ_ACCESSOR_PROPERTY("Resource", GetResourceFile, SetResourceFile)->AddAttributes(new ezAssetBrowserAttribute("CompatibleAsset_StateMachine")),
+    EZ_ACCESSOR_PROPERTY("InitialState", GetInitialState, SetInitialState),
+    EZ_MEMBER_PROPERTY("ResetOnEnter", m_bResetOnEnter)->AddAttributes(new ezDefaultValueAttribute(true)),
+    EZ_MEMBER_PROPERTY("ResetOnExit", m_bResetOnExit)->AddAttributes(new ezDefaultValueAttribute(true)),
+  }
+  EZ_END_PROPERTIES;
+  EZ_BEGIN_ATTRIBUTES
+  {
+    EZ_INSTANCE_DATA_TYPE_ATTRIBUTE(ezStateMachineState_NestedStateMachine::InstanceData),
+  }
+  EZ_END_ATTRIBUTES;
+}
+EZ_END_DYNAMIC_REFLECTED_TYPE;
+// clang-format on
+
+ezStateMachineState_NestedStateMachine::ezStateMachineState_NestedStateMachine(const char* szName /*= nullptr*/)
+  : ezStateMachineState(szName)
+{
+}
+
+ezStateMachineState_NestedStateMachine::~ezStateMachineState_NestedStateMachine() = default;
+
+void ezStateMachineState_NestedStateMachine::OnEnter(ezStateMachineInstance& instance, void* pStateInstanceData) const
+{
+  auto& pStateMachineInstance = static_cast<InstanceData*>(pStateInstanceData)->m_pStateMachineInstance;
+
+  bool bSetInitialState = m_bResetOnEnter;
+
+  if (pStateMachineInstance == nullptr)
+  {
+    if (m_hResource.IsValid() == false)
+      return;
+
+    ezResourceLock<ezStateMachineResource> pStateMachineResource(m_hResource, ezResourceAcquireMode::BlockTillLoaded_NeverFail);
+    if (pStateMachineResource.GetAcquireResult() != ezResourceAcquireResult::Final)
+    {
+      ezLog::Error("Failed to load state machine '{}'", GetResourceFile());
+      return;
+    }
+
+    pStateMachineInstance = pStateMachineResource->CreateInstance(instance.GetOwner());
+    pStateMachineInstance->SetBlackboard(instance.GetBlackboard());
+    bSetInitialState = true;
+  }
+
+  if (bSetInitialState)
+  {
+    pStateMachineInstance->SetStateOrFallback(m_sInitialState).IgnoreResult();
+  }
+}
+
+void ezStateMachineState_NestedStateMachine::OnExit(ezStateMachineInstance& instance, void* pStateInstanceData) const
+{
+  if (m_bResetOnExit)
```

Comment:
Not sure how reset on enter and exit make a difference? I will always exit before I enter again, so whether I reset the start state on enter or on exit, where is the difference?

Reply:
Made it one bool flag and renamed it. Also added comment to clarify what it does.